### ISSUE TITLE
Building error

### ISSUE THREAD

**Issue by andrewjrobinson** [TGAC/KAT#10]
Jellyfish that comes with KAT is having problems building. It appears something went wrong with libtool but I am not sure: 

```
./libtool: line 1901: 3: command not found
```

**Downloaded**: https://github.com/TGAC/KAT/releases/download/Release-2.0.3/kat-2.0.3.tar.gz

**Using versions**:
- OS: RHEL6.6
- gcc/4.8.2
- boost-gcc/1.55.0
- libtool/2.4.6

**Make output**:
```
$ make
make  all-recursive
make[1]: Entering directory `/usr/local/src/kat/kat-2.0.3'
Making all in deps/jellyfish-2.2.0
make[2]: Entering directory `/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0'
make  all-recursive
make[3]: Entering directory `/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0'
Making all in .
make[4]: Entering directory `/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0'
  CXX      lib/rectangular_binary_matrix.lo
./libtool: line 1901: 3: command not found
  CXX      lib/mer_dna.lo
./libtool: line 1901: 3: command not found
  CXX      lib/storage.lo
./libtool: line 1901: 3: command not found
  CXX      lib/allocators_mmap.lo
./libtool: line 1901: 3: command not found
  CXX      lib/misc.lo
./libtool: line 1901: 3: command not found
  CXX      lib/int128.lo
./libtool: line 1901: 3: command not found
  CXX      lib/thread_exec.lo
./libtool: line 1901: 3: command not found
  CXX      lib/jsoncpp.lo
./libtool: line 1901: 3: command not found
  CXX      lib/time.lo
./libtool: line 1901: 3: command not found
  CXX      lib/generator_manager.lo
./libtool: line 1901: 3: command not found
  CXXLD    libjellyfish-2.0.la
libtool: link: `lib/rectangular_binary_matrix.lo' is not a valid libtool object
make[4]: *** [libjellyfish-2.0.la] Error 1
make[4]: Leaving directory `/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0'
make[3]: *** [all-recursive] Error 1
make[3]: Leaving directory `/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0'
make[2]: *** [all] Error 2
make[2]: Leaving directory `/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/usr/local/src/kat/kat-2.0.3'
make: *** [all] Error 2
```

**Configure output**:
```
$ export VERSION=2.0.3
$ ./configure --prefix=/usr/local/kat/$VERSION-gcc --with-boost=$BOOST_DIR
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether UID '2504' is supported by ustar format... yes
checking whether GID '2504' is supported by ustar format... yes
checking how to create a ustar tar archive... gnutar
checking for style of include used by make... GNU
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... gcc3
checking how to run the C++ preprocessor... g++ -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether g++ accepts -g... (cached) yes
checking dependency style of g++... (cached) gcc3
checking whether g++ supports C++11 features by default... no
checking whether g++ supports C++11 features with -std=gnu++11... yes
checking for gcc... gcc
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for ar... ar
checking the archiver (ar) interface... ar
checking for error_at_line... yes
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking whether strerror_r is declared... yes
checking for strerror_r... yes
checking whether strerror_r returns char *... yes
checking for int16_t... yes
checking for int32_t... yes
checking for int64_t... yes
checking for uint8_t... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking for uint64_t... yes
checking for size_t... yes
checking for floor... yes
checking for memset... yes
checking for sqrt... yes
checking for inline... inline
checking for string.h... (cached) yes
checking iostream usability... yes
checking iostream presence... yes
checking for iostream... yes
checking stdio.h usability... yes
checking stdio.h presence... yes
checking for stdio.h... yes
checking fstream usability... yes
checking fstream presence... yes
checking for fstream... yes
checking memory usability... yes
checking memory presence... yes
checking for memory... yes
checking vector usability... yes
checking vector presence... yes
checking for vector... yes
checking glob.h usability... yes
checking glob.h presence... yes
checking for glob.h... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking if compiler needs -Werror to reject unknown flags... no
checking for the pthreads library -lpthreads... no
checking whether pthreads work without any flags... no
checking whether pthreads work with -Kthread... no
checking whether pthreads work with -kthread... no
checking for the pthreads library -llthread... no
checking whether pthreads work with -pthread... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking if more special flags are required for pthreads... no
checking for PTHREAD_PRIO_INHERIT... no
checking for gnuplot... no

checking for boostlib >= 1.52... yes
checking whether the Boost::Filesystem library is available... yes
checking for exit in -lboost_filesystem... yes
checking for /usr/local/boost/1.55.0-gcc//lib/libboost_filesystem.a... yes
checking whether the Boost::System library is available... yes
checking for exit in -lboost_system... yes
checking for /usr/local/boost/1.55.0-gcc//lib/libboost_system.a... yes
checking whether the Boost::Program_Options library is available... yes
checking for exit in -lboost_program_options... yes
checking for /usr/local/boost/1.55.0-gcc//lib/libboost_program_options.a... yes
checking whether the Boost::Chrono library is available... yes
checking for exit in -lboost_chrono... yes
checking for /usr/local/boost/1.55.0-gcc//lib/libboost_chrono.a... yes
checking whether the Boost::Timer library is available... yes
checking for exit in -lboost_timer... yes
checking for /usr/local/boost/1.55.0-gcc//lib/libboost_timer.a... yes
checking whether the Boost::Unit_Test_Framework library is available... yes
checking for exit in -lboost_unit_test_framework... yes
checking for /usr/local/boost/1.55.0-gcc//lib/libboost_unit_test_framework.a... yes
configure: WARNING: cache variable ac_cv_prog_gnuplot_found contains a newline
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating tests/Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
=== configuring in deps/jellyfish-2.2.0 (/usr/local/src/kat/kat-2.0.3/deps/jellyfish-2.2.0)
configure: running /bin/sh ./configure --disable-option-checking '--prefix=/usr/local/kat/2.0.3-gcc'  '--with-boost=/usr/local/boost/1.55.0-gcc/' 'LDFLAGS=-L/usr/local/libtool/2.4.6/lib -L/usr/local/boost/1.55.0-gcc/lib -L/usr/local/gcc/4.8.2/lib64 -L/usr/local/gcc/4.8.2/lib -L/usr/local/mpc/1.0.2/lib -L/usr/local/mpfr/3.1.2/lib -L/usr/local/gmp/5.1.3/lib' 'CPPFLAGS=-I/usr/local/libtool/2.4.6/include -I/usr/local/boost/1.55.0-gcc/include -I/usr/local/gcc/4.8.2/include -I/usr/local/mpc/1.0.2/include -I/usr/local/mpfr/3.1.2/include -I/usr/local/gmp/5.1.3/include' --cache-file=/dev/null --srcdir=.
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking how to print strings... printf
checking for style of include used by make... GNU
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for a sed that does not truncate output... /bin/sed
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for fgrep... /bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1966080
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... gcc3
checking how to run the C++ preprocessor... g++ -E
checking for ld used by g++... /usr/bin/ld -m elf_x86_64
checking if the linker (/usr/bin/ld -m elf_x86_64) is GNU ld... yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC -DPIC
checking if g++ PIC flag -fPIC -DPIC works... yes
checking if g++ static flag -static works... no
checking if g++ supports -c -o file.o... yes
checking if g++ supports -c -o file.o... (cached) yes
checking whether the g++ linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for md5sum... md5sum
checking for yaggo... false
checking for __int128... yes
checking for std::numeric_limits<__int128>... no
checking for _NSGetExecutablePath... no
checking for execinfo.h... yes
checking for ext/stdio_filebuf.h... yes
checking for siginfo_t.si_int... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tests/compat.sh
config.status: creating jellyfish-2.0.pc
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
configure: WARNING: cache variable ac_cv_prog_gnuplot_found contains a newline
```

**Comment by maplesond** [comment]
Hi Andrew, thanks for the raising this and providing the detailed logs.  I'll look into this today and get back to you.  In the meantime you could try cloning directly from github and see if you can install that way.

Best regards,
Dan Mapleson

**Comment by maplesond** [comment]
I've taken a look on my build machines and all seem to compile fine (both from distributable and git clone), although I don't have quite the same environment setup:

Desktop (s):
OS: Mint 17.2
GCC: 4.8.4
Boost: 1.55.0
libtool: 2.4.2

Cluster:
OS: CentOS release 6.7
GCC: 4.9.1
Boost: 1.59.0
libtool: 2.4.2

Out of interest have you tried installing any earlier versions of KAT?  If so did they succeed or fail?  Also could you send me your config.log files.  Both for jellyfish (./deps/jellyfish-2.2.0/config.log) and kat (./config.log) please.  This might help me work out what's happening on your system.

As an aside, I noticed that there was a check failing for gnuplot, which I've now fixed and is currently head of the develop branch. This shouldn't affect your main problem however.

**Comment by maplesond** [comment]
Another thing to check for is whether or not libtool (./deps/jellyfish-2.2.0/libtool) is created by the configure script.  It should be there after untarring the distributable (or cloning the repo), but should be after running the configure script.  

If this doesn't work, one thing to try would be modifying how jellyfish calls libtool.  To do this you'll need to work from the cloned repo (not the distributable).  Firstly, open the jellyfish configure.ac (./deps/jellyfish-2.2.0/configure.ac) file.  Then change line 8 from "AC_PROG_LIBTOOL" to "LT_INIT". (AC_PROG_LIBTOOL is a deprecated way of initialising libtool, LT_INIT is the currently recommended way).  Then go back to the kat root dir, run "./autogen.sh", then do the "./configure".  Then check to see if the "libtool" script has been made in the jellyfish dir.  If it has then hopefully make will work properly now.

**Comment by andrewjrobinson** [comment]
Thanks for all the suggestions Daniel,

The libtool file didn't exist in the repo versions of source (before running ./autogen.sh) but did in release versions (and after running ./autogen.sh on repo versions) so appears to be all ok.

I have tried:
1) Changing the LT_INIT line: same error
1) Installing from latest head of repo: same error
2) Installing an older version (2.0.1): same error
3) Installing with an older version of libtool (2.4.2): same error

I had already installed jellyfish v2.2.3 on this system and it worked so I tried installing jellyfish v2.2.0 which showed the similar errors to what I was getting with the KAT version of jellyfish.  So the error seems to be with the jellyfish v2.2.0 on my system.

I should have mentioned that this is a HPC and uses "Environment Modules" to bring particular versions of software into users PATH (that are installed in NON-standard prefixes).

I tried to upgrade the version of jellyfish in the latest head of KAT but was getting errors with swig and now it wants valgrind.  Do you do any alterations to jellyfish that you embed? and is it possible to tell it to use and external version of jellyfish (from the PATH)?

**Comment by maplesond** [comment]
Hi Andrew, thanks for the feedback.  It sounds like the jellyfish authors have fixed a few issues and made the software more portable since v2.2.0.  I'll update the jellyfish version in KAT for a future release.  

With regard to jellyfish modifications, I have made some small changes in the jellyfish configure.ac file, to remove some checks which are unnecessary for use within KAT (the swig and valgrind checks for example).  

For KAT V2 we made the decision to embed jellyfish within KAT, and therefore prevent version mismatches, as we think this improves stability of the system.  Previously we were running into issues where hashes made with different versions of jellyfish were not compatible (even from dot releases), so embedding jellyfish avoids that problem completely.  Also we make extensive use of the jellyfish library's API, and would like to avoid and potential issues when changes to the API occur.  So unfortunately there's noway of overriding the embedded version of jellyfish in KAT.

Hope that helps.  I'll leave the ticket open for now.  I'll close it when I've embedded a newer version of jellyfish.
